internal/storage/metricstore/elasticsearch/reader.go

```diff
@@ -185,24 +205,23 @@ func (r MetricsReader) buildQuery(params metricstore.BaseQueryParameters, timeRa
 	return *boolQuery
 }
 
-// processCallRateMetrics processes the MetricFamily to calculate call rates
-func getCallRateProcessMetrics(mf *metrics.MetricFamily, params metricstore.BaseQueryParameters) *metrics.MetricFamily {
-	lookback := int(math.Ceil(float64(params.RatePer.Milliseconds()) / float64(params.Step.Milliseconds())))
-	if lookback < 1 {
-		lookback = 1
-	}
-
+// processMovingFunction applies a given processing function over a moving window of metric points.
+// This is the core generic function that contains the shared logic.
+func processMovingFunction(mf *metrics.MetricFamily, lookback int, processor func(window []*metrics.MetricPoint) float64) *metrics.MetricFamily {
 	for _, metric := range mf.Metrics {
 		points := metric.MetricPoints
-		var processedPoints []*metrics.MetricPoint
+		if len(points) == 0 {
+			continue
+		}
+
+		processedPoints := make([]*metrics.MetricPoint, 0, len(points))
 
-		for i := range points {
-			currentPoint := points[i]
+		for i, currentPoint := range points {
 			currentValue := currentPoint.GetGaugeValue().GetDoubleValue()
 
 			// Elasticsearch's percentiles aggregation returns 0.0 for time buckets with no documents
 			// These aren't true zero values but represent missing data points in sparse time series
-			// We convert them to NaN to distinguish from actual measured zero values (slope of 0)
+			// We convert them to NaN to distinguish from actual measured zero values
 			if currentValue == 0.0 {
 				processedPoints = append(processedPoints, &metrics.MetricPoint{
```

Comment:
I don't follow this. There are two distinct functions being performed - one is verifying that the interval has a real value (or NaN). Even if it's NaN it only applies to the _current interval_ not to the output of the sliding window aggregation, which is the second function. Wouldn't the correct logic be
  1. points = raw_points.map(p => maybeNaN(p)
  2. now apply sliding window func over points